File: XcodeBetterRefactorTools/Xcode Additions/XMASFakeProtocolPersister.m
#import "XMASFakeProtocolPersister.h"
#import "SwiftCompatibilityHeader.h"

@interface XMASFakeProtocolPersister ()
@property (nonatomic, strong) NSFileManager *fileManager;
@property (nonatomic, strong) XMASSwiftProtocolFaker *protocolFaker;
@end

@implementation XMASFakeProtocolPersister

- (instancetype)initWithProtocolFaker:(XMASSwiftProtocolFaker *)protocolFaker
                          fileManager:(NSFileManager *)fileManager {
    if (self = [super init]) {
        self.fileManager = fileManager;
        self.protocolFaker = protocolFaker;
    }

    return self;
}

- (void)persistFakeForProtocol:(ProtocolDeclaration *)protocolDecl
                nearSourceFile:(NSString *)sourceFilePath {
    // get dir for source file path
    NSString *dirContainingSource = [sourceFilePath stringByDeletingLastPathComponent];
    NSString *fakesDir = [dirContainingSource stringByAppendingPathComponent:@"fakes"];

    // create ^^/fakes if it does not exist
    BOOL dirExists = [self.fileManager fileExistsAtPath:fakesDir isDirectory:nil];
    if (!dirExists) {
        [self.fileManager createDirectoryAtPath:fakesDir withIntermediateDirectories:YES attributes:nil error:nil];
    }

    // create fake for protocol
    NSError *error = nil;
    NSData *fileContents = [[self.protocolFaker fakeForProtocol:protocolDecl error:&error] dataUsingEncoding:NSUTF8StringEncoding];
    if (error != nil) {
        NSString *failureReason = error.localizedFailureReason;
        [[NSException exceptionWithName:@"" reason:failureReason userInfo:nil] raise];
    }

    NSString *fakeFileName = [[@"Fake" stringByAppendingString:protocolDecl.name] stringByAppendingString:@".swift"];
    NSString *pathToFake = [fakesDir stringByAppendingPathComponent:fakeFileName];

    // write out file contents into ^^/fakes/fake_blah_blah_blah.swift
    [self.fileManager createFileAtPath:pathToFake
                              contents:fileContents
                            attributes:nil];
}

#pragma mark - NSObject

- (instancetype)init {
    [self doesNotRecognizeSelector:_cmd];
    return nil;
}

@end

File: BetterRefactorToolsKit/Use Cases/Parse Swift Protocol/XMASFakeProtocolPersister.swift
import Foundation

@objc public protocol XMASFileManager {
    func fileExistsAtPath(
        path: String,
        isDirectory: UnsafeMutablePointer<ObjCBool>) -> Bool

    func createDirectoryAtPath(
        path: String,
        withIntermediateDirectories: Bool,
        attributes: [String : AnyObject]?) throws

    func createFileAtPath(
        path: String,
        contents: NSData?,
        attributes: [String : AnyObject]?) -> Bool
}

extension NSFileManager : XMASFileManager {

}

@objc public class XMASFakeProtocolPersister : NSObject {
    private(set) public var fileManager : XMASFileManager
    private(set) public var protocolFaker : XMASSwiftProtocolFaking

    @objc public init(protocolFaker: XMASSwiftProtocolFaking, fileManager: XMASFileManager) {
        self.fileManager = fileManager
        self.protocolFaker = protocolFaker
    }

    @objc public func persistFakeForProtocol(
        protocolDecl : ProtocolDeclaration,
        nearSourceFile: String) throws
    {
        let dirContainingSource = (nearSourceFile as NSString).stringByDeletingLastPathComponent
        let fakesDir = (dirContainingSource as NSString).stringByAppendingPathComponent("fakes")

        let fakeFileName : String = "Fake".stringByAppendingString(protocolDecl.name).stringByAppendingString(".swift")
        let pathToFake = (fakesDir as NSString).stringByAppendingPathComponent(fakeFileName)

        if !fileManager.fileExistsAtPath(fakesDir as String, isDirectory: nil) {
            try self.fileManager.createDirectoryAtPath(
                fakesDir,
                withIntermediateDirectories: true,
                attributes: nil)
        }

        do {
            let fileContents = try self.protocolFaker.fakeForProtocol(protocolDecl)

            let fileData : NSData = fileContents.dataUsingEncoding(NSUTF8StringEncoding)!
            fileManager.createFileAtPath(pathToFake, contents: fileData, attributes: nil)

        } catch let error as NSError {
            throw error
        }
    }
}